packages/connect-core/src/async-iterable.ts

```diff
@@ -1126,3 +1126,136 @@ export function makeIterableAbortable<T>(
     },
   };
 }
+
+// A class representing a reader that reads from the given AsyncIterable.
+export class Reader<T> {
+  private it: AsyncIterable<T>;
+
+  constructor(it: AsyncIterable<T>) {
+    this.it = it;
+  }
+  [Symbol.asyncIterator](): AsyncIterator<T> {
+    return {
+      next: async () => {
+        for await (const item of this.it) {
+          return { value: item };
+        }
+        return { value: "", done: true };
+      },
+    };
+  }
+}
+
+// QueueElement represents an element in the writer queue, which consists of the payload being written as well as an
+// associated resolve function to be invoked/resolved when the written element is read from the queue via the async
+// iterator.
+interface QueueElement<T> {
+  payload: IteratorResult<T>;
+  resolve?: () => void;
+  reject?: (reason?: Error) => void;
+}
+
+// WritableIterable represents an AsyncIterable that is able to be written to.
+export interface WritableIterable<T> extends AsyncIterable<T> {
+  write: (payload: T) => Promise<void>;
+  close: () => Promise<void>;
+  isClosed: () => boolean;
+}
+
+// Create an instance of a WritableIterable of type T
+export function createWritableIterable<T>(): WritableIterable<T> {
+  let queue: QueueElement<T>[] = [];
+  // Represents the resolve function of the promise returned by the async iterator if no values exist in the queue at
+  // the time of request.  It is resolved when a value is successfully received into the queue.
+  let queueResolve: ((val: IteratorResult<T>) => void) | undefined;
+  let error: Error | undefined = undefined;
+
+  const process = async (payload: IteratorResult<T, undefined>) => {
+    // // If the writer's internal error was set, then reject any attempts at processing a payload.
+    if (error) {
+      return Promise.reject(String(error));
+    }
+    // If there is an iterator resolver then a consumer of the async iterator is waiting on a value.  So resolve that
+    // promise with the new value being sent and return a promise that is immediately resolved
+    if (queueResolve) {
+      queueResolve(payload);
+      queueResolve = undefined;
+      return Promise.resolve();
+    }
+    const elem: QueueElement<T> = {
+      payload,
+    };
+    const prom = new Promise<void>((resolve, reject) => {
+      elem.resolve = resolve;
+      elem.reject = reject;
+    });
+    // Otherwise no one is waiting on a value yet so add it to the queue and return a promise that will be resolved
+    // when someone reads this value
+    queue.push(elem);
+
+    return prom;
+  };
+  const send = async (payload: T) => {
+    return process({ value: payload, done: false });
+  };
+  const close = async () => {
+    return process({ value: undefined, done: true });
+  };
```

Comment:
These two functions are only called in the corresponding `write` and `close` methods - can't we simply call `process` there?

This works the same, but doesn't need `const close ...`:

```ts
    async close() {
      if (closed) {
        throw new ConnectError("cannot close, already closed");
      }
      closed = true;
      return process({ value: undefined, done: true });
    },
```